### PR TITLE
Add DEBUGGING.md

### LINKED ISSUE #710: Document Debugging OpenTofu
### OpenTofu Version

```shell
OpenTofu v1.6.0-alpha2
on darwin_arm64
```

### Use Cases

Debugging is one of the crucial ways to develop code, and there are various ways to debug OpenTofu. Having a dedicated page for debugging (for example a `DEBUGGING.md` file) would help the developers, contributors to better engage with OpenTofu code and simplify the development/bug-fixing process. 

### Attempted Solutions

-

### Proposal

The proposal for having a debugging guide came from @Yantrio [in this slack thread](https://opentofucommunity.slack.com/archives/C05RKE7PS72/p1697026244736189?thread_ts=1697025755.016759&cid=C05RKE7PS72):
> Overall, I think it may be good that we have one guide about getting started with debugging, and then people can work from there.

Also mentioned in the same thread, there is #667 for adding vscode files for debugging. This issue is intended to replace it and cover a larger solution for documenting the debugging options.   

### References

Slack thread for the discussion: https://opentofucommunity.slack.com/archives/C05RKE7PS72/p1697025755016759
Related GH Issues for the vscode files: #667 and the PR #668

### PR DESCRIPTION
**Pull request by serdardalgic**
resolves #710.

I'd be glad if you can add a hacktoberfest-accepted tag to this PR, please.

A note to the reviewers, English is not my native language, so pardon me for any grammatical mistakes, and please don't hesitate to suggest a clearer version of the sentences.  

I'm not familiar with Goland IDE, so maybe @kislerdm might be interested in contributing to this guide if necessary for including Goland IDE debugging information. :)

Please review.

## Target Release

1.6.0

### NEW FEATURES | UPGRADE NOTES | ENHANCEMENTS | BUG FIXES | EXPERIMENTS

 

-

### CODE REVIEW

File: DEBUGGING.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+- [Using the debug-opentofu script](#using-the-debug-opentofu-script)
+- [Using spew](#using-spew)
+- [Using VsCode](#using-vscode)
+
+<!-- /TOC -->
+
+## Using the debug-opentofu script
+
+[debug-opentofu](./scripts/debug-opentofu) is a helper script to launch OpenTofu inside the "dlv" debugger, configured to await a remote debugging connection on port 2345. For more details on how to use this script, please refer to the documentation at the beginning of this script.
```
Comment (kislerdm) [inline DEBUGGING.md:19]: **(suggestion, non-blocking)** Hey! WDYT about linking [delve](https://github.com/go-delve/delve/tree/master)?

**(suggestion, nitpick)** Also, it makes sense to replace the link in the script's [comments](https://github.com/opentofu/opentofu/blob/03c8f6cebd2dcca4b8a1c6439322cc02395d46c1/scripts/debug-opentofu#L10):

_From_:
```
#    https://github.com/derekparker/delve/tree/master/Documentation/installation
```

_To_:
```
#    https://github.com/go-delve/delve/tree/master/Documentation/installation
```

WDYT?

Comment (serdardalgic) [reply DEBUGGING.md:19]: Sure, both are good suggestions, thanks :)

Comment (serdardalgic) [reply DEBUGGING.md]: Both suggestions are applied, resolving the conversation. Please don't hesitate to reopen if I missed anything.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -14,14 +14,15 @@ Generally, we appreciate external contributions very much and would love to work
 
 ---
 
-<!-- MarkdownTOC autolink="true" -->
+<!-- TOC -->
```
Comment (Yantrio) [inline CONTRIBUTING.md:17]: I think these headers were there for a reason for auto-generation of TOC. Maybe we need to settle on a tool to generate these, but for now I think it's best if we revert this small change.

File: DEBUGGING.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+<!-- /TOC -->
+
+## Using the debug-opentofu script
+
+[debug-opentofu](./scripts/debug-opentofu) is a helper script to launch OpenTofu inside the ["dlv" debugger](https://github.com/go-delve/delve), configured to await a remote debugging connection on port 2345. For more details on how to use this script, please refer to the documentation at the beginning of this script.
+
+## Using spew
+
+[Go-spew](https://github.com/davecgh/go-spew) implements a deep pretty printer for Go data structures to aid in debugging. If you prefer to use println debugging, spew.Dump might be helpful.
```
Comment (Yantrio) [inline DEBUGGING.md:23]:
```suggestion
[Go-spew](https://github.com/davecgh/go-spew) implements a deep pretty printer for Go data structures to aid in debugging. If you prefer to use println debugging, `spew.Dump` might be helpful.
```

**Review by Yantrio — CHANGES_REQUESTED** [review]
Just a couple nitpicks :)

File: CONTRIBUTING.md (hunk 2 above)
Comment (serdardalgic) [reply CONTRIBUTING.md]: Sure, I guess that must be [the Sublime Plugin](https://packagecontrol.io/packages/MarkdownTOC) to generate the MarkdownTOC, as the header suggests. 

I don't have any preference, TBH. I favor any automated solution that makes it easier to traverse the documentation and requires minimal effort to maintain.  👍

File: DEBUGGING.md (hunk 3 above)
Comment (serdardalgic) [reply DEBUGGING.md]: Direct commits from GitHub UI doesn't comply with DCO, so I `commit --amend --sign-off` this commit, FYI.

File: DEBUGGING.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# Debugging the OpenTofu code
+
+There are various ways to debug the OpenTofu code. There is no ultimate "right" answer, this document intends to collect some of those ways. The order of debugging techniques is completely random.
+
+If you would like to contribute to this debugging guide, please [create a GitHub issue](https://github.com/opentofu/opentofu/issues/new/choose) and propose the enhancement. After that, you can create a pull request and reference this issue in your PR.
+
+For further information on contributing to the code, please refer to the [CONTRIBUTING.md](./CONTRIBUTING.md) file.
+
+<!-- TOC -->
```
Comment (serdardalgic) [inline DEBUGGING.md:9]: @Yantrio I'd like to keep this header as `TOC` instead of `MarkdownTOC` if there is no objection. (=

**Comment by serdardalgic** [comment]
Ping again, as I don't have anything else to add to this PR. @Yantrio @cube2222, can you review at your convenience, please? Thank you.

**Review by Yantrio — APPROVED** [review]

File: CONTRIBUTING.md (hunk 2 above)
Comment (serdardalgic) [reply CONTRIBUTING.md]: I've reverted this change, the conversation is marked as outdated, but not resolved, I don't know why. 
I'm resolving this conversation.

**Comment by cube2222** [comment]
Thanks for the contribution @serdardalgic!

**Pull request merged**